icon.c

```diff
@@ -28,7 +37,144 @@ static double fit_to_square(int width, int height, int square_size) {
 	return longest > square_size ? square_size/longest : 1.0;
 }
 
-struct mako_icon *create_icon(const char *path, double max_size) {
+// Attempt to find a full path for a notification's icon_name, which may be:
+// - An absolute path, which will simply be returned (as a new string)
+// - A file:// URI, which will be converted to an absolute path
+// - A Freedesktop icon name, which will be resolved within the configured
+//   `icon-path` using something that looks vaguely like the algorithm defined
+//   in the icon theme spec (https://standards.freedesktop.org/icon-theme-spec/icon-theme-spec-latest.html)
+//
+// Returns the resolved path, or NULL if it was unable to find an icon. The
+// return value must be freed by the caller.
+static char *resolve_icon(struct mako_notification *notif) {
+	char *icon_name = notif->app_icon;
+	if (icon_name[0] == '\0') {
+		return NULL;
+	}
+	if (icon_name[0] == '/' || icon_name[0] == '~') {
+		return strdup(icon_name);
+	}
+	if (strstr(icon_name, "file://") == icon_name) {
+		// Just chop off the scheme.
+		return strdup(icon_name + 7);
+	}
+
+	// Determine the largest scale factor of any attached output.
+	int32_t max_scale = 1;
+	struct mako_output *output = NULL;
+	wl_list_for_each(output, &notif->state->outputs, link) {
+		if (output->scale > max_scale) {
+			max_scale = output->scale;
+		}
+	}
+
+	static const char *fallback = "%s:/usr/share/icons/hicolor";
+	char *search = mako_asprintf(fallback, notif->style.icon_path);
+
+	char *saveptr = NULL;
+	char *theme_path = strtok_r(search, ":", &saveptr);
+
+	// Match all icon files underneath of the theme_path followed by any icon
+	// size and category subdirectories. This pattern assumes that all the
+	// files in the icon path are valid icon types.
+	static const char *pattern_fmt = "%s/*/*/%s.*";
+
+	char *icon_path = NULL;
+	int32_t last_icon_size = 0;
+	while (theme_path) {
```

Comment:
We also probably want to skip empty `theme_path`s (this happens if `notif->style.icon_path` is empty for instance)